flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java

```diff
@@ -202,13 +204,53 @@ public void run() {
 				}
 			}
 
-			// seek the consumer to the initial offsets
+			List<KafkaTopicPartition> partitionsWithNoOffset = new ArrayList<>();
 			for (KafkaTopicPartitionState<TopicPartition> partition : subscribedPartitions()) {
-				if (partition.isOffsetDefined()) {
+				if (!partition.isOffsetDefined()) {
+					partitionsWithNoOffset.add(partition.getKafkaTopicPartition());
+				} else {
 					consumer.seek(partition.getKafkaPartitionHandle(), partition.getOffset() + 1);
 				}
 			}
 
+			if (partitionsWithNoOffset.size() == subscribedPartitions().length) {
+				// if all partitions have no initial offsets, that means we're starting fresh
+				switch (startupMode) {
+					case EARLIEST:
+						LOG.info("Setting starting point as earliest offset for partitions {}", partitionsWithNoOffset);
+
+						for (KafkaTopicPartitionState<TopicPartition> partition : subscribedPartitions()) {
+							consumer.seekToBeginning(partition.getKafkaPartitionHandle());
+						}
+						break;
+					case LATEST:
+						LOG.info("Setting starting point as latest offset for partitions {}", partitionsWithNoOffset);
+
+						for (KafkaTopicPartitionState<TopicPartition> partition : subscribedPartitions()) {
+							consumer.seekToEnd(partition.getKafkaPartitionHandle());
+						}
+						break;
+					default:
+					case GROUP_OFFSETS:
+						LOG.info("Using group offsets in Kafka of group.id {} as starting point for partitions {}",
+							kafkaProperties.getProperty("group.id"), partitionsWithNoOffset);
+						// don't need to do anything; the KafkaConsumer by default finds group offsets from Kafka brokers
+				}
+			} else if (partitionsWithNoOffset.size() > 0 && partitionsWithNoOffset.size() < subscribedPartitions().length) {
```

Comment:
@rmetzger after looking at https://issues.apache.org/jira/browse/FLINK-3037, do you think the proposed changes here (in the `Kafka09Fetcher`, overall) actually fixes that issue?